--- tigervnc-1.0.90/unix/xserver/hw/vnc/Makefile.am~	2011-04-20 13:46:16.993047050 +0200
+++ tigervnc-1.0.90/unix/xserver/hw/vnc/Makefile.am	2011-04-20 14:03:35.942724608 +0200
@@ -1,11 +1,12 @@
 TIGERVNC_SRCDIR=${top_srcdir}/../..
 LIB_DIR=${top_builddir}/../../common
 
+OS_LIB=$(LIB_DIR)/os/libos.a
 RFB_LIB=$(LIB_DIR)/rfb/librfb.la
 RDR_LIB=$(LIB_DIR)/rdr/librdr.la
 NETWORK_LIB=$(LIB_DIR)/network/libnetwork.la
 XREGION_LIB=$(LIB_DIR)/Xregion/libXregion.la
-COMMON_LIBS=$(NETWORK_LIB) $(RFB_LIB) $(RDR_LIB) $(XREGION_LIB)
+COMMON_LIBS=$(NETWORK_LIB) $(OS_LIB) $(RFB_LIB) $(RDR_LIB) $(XREGION_LIB)
 
 noinst_LTLIBRARIES = libvnccommon.la
 
--- tigervnc-1.2.0/CMakeLists.txt~	2012-03-09 21:34:29.000000000 +0100
+++ tigervnc-1.2.0/CMakeLists.txt	2012-03-23 13:47:57.895270032 +0100
@@ -316,7 +316,7 @@
   if(APPLE)
     set(FLTK_LIBRARIES "-framework Carbon -framework Cocoa -framework ApplicationServices")
   elseif(NOT WIN32)
-    set(FLTK_LIBRARIES "-ldl")
+    set(FLTK_LIBRARIES "-ldl -lfontconfig")
   endif()
   message(STATUS "Using included FLTK library")
 endif()
