diff --git a/unix/vncserver b/unix/vncserver
index 68be032..ac0e993 100755
--- a/unix/vncserver
+++ b/unix/vncserver
@@ -678,6 +678,7 @@ sub Usage
 	"                 [-geometry <width>x<height>]\n".
 	"                 [-pixelformat rgbNNN|bgrNNN]\n".
 	"                 [-fp <font-path>]\n".
+	"                 [-cc <visual>]\n".
 	"                 [-fg]\n".
 	"                 [-autokill]\n".
 	"                 [-noxstartup]\n".
