--- tigervnc-1.12.0/unix/xserver/hw/vnc/xorg-version.h.orig	2022-07-14 14:47:19.295275332 +0200
+++ tigervnc-1.12.0/unix/xserver/hw/vnc/xorg-version.h	2022-07-14 16:04:12.530283292 +0200
@@ -33,8 +33,8 @@
 #error "X.Org older than 1.16 is not supported"
 #endif
 
-#if XORG_AT_LEAST(1, 21, 0)
-#error "X.Org newer than 1.20 is not supported"
+#if XORG_AT_LEAST(1, 22, 0)
+#error "X.Org newer than 1.21 is not supported"
 #endif
 
 #endif
--- tigervnc-1.12.0/unix/xserver/hw/vnc/xvnc.c.orig	2021-11-09 08:51:28.000000000 +0100
+++ tigervnc-1.12.0/unix/xserver/hw/vnc/xvnc.c	2022-07-14 16:04:32.873506417 +0200
@@ -69,7 +69,14 @@ extern char buildtime[];
 #undef VENDOR_RELEASE
 #undef VENDOR_STRING
 #include "version-config.h"
-#include "site.h"
+
+/* Default logging parameters. */
+#ifndef DEFAULT_LOG_VERBOSITY
+#define DEFAULT_LOG_VERBOSITY		0
+#endif
+#ifndef DEFAULT_LOG_FILE_VERBOSITY
+#define DEFAULT_LOG_FILE_VERBOSITY	3
+#endif
 
 #define XVNCVERSION "TigerVNC 1.12.0"
 #define XVNCCOPYRIGHT ("Copyright (C) 1999-2021 TigerVNC Team and many others (see README.rst)\n" \
@@ -117,8 +124,7 @@ static void
 vncPrintBanner(void)
 {
     ErrorF("\nXvnc %s - built %s\n%s", XVNCVERSION, buildtime, XVNCCOPYRIGHT);
-    ErrorF("Underlying X server release %d, %s\n\n", VENDOR_RELEASE,
-           VENDOR_STRING);
+    ErrorF("Underlying X server release %d\n\n", VENDOR_RELEASE);
 }
 
 static void
